File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int luhnsAlgorithm(long unsigned int number);
int getSize(long unsigned int number);
int size = 0;

int main()
{
    long unsigned int card_number = 0;
    int flag = 0;
    card_number = get_long("Number: ");
    getSize(card_number);

    if ((size != 15 && size != 16 && size != 13) || card_number == 0)
    {
        printf("INVALID\n");
        return 0;
    }
    else
    {
        if (size == 13)
        {
            if (card_number / 1000000000000 == 4)
            {
                flag = 3;
            }
        }
        if (size == 15)
        {
            if ((card_number / 10000000000000 == 34) || (card_number / 10000000000000 == 37))
            {
                flag = 1;
            }
        }
        if (size == 16)
        {
            if (card_number / 1000000000000000 == 4)
            {
                flag = 3;
            }
            else
            {
                if ((card_number / 100000000000000 >= 51) && (card_number / 100000000000000 <= 55))
                {
                    flag = 2;
                }
            }
        }

        int total = luhnsAlgorithm(card_number);
        while (card_number > 0)
        {
            total = total + (card_number % 10);
            card_number = card_number / 100;
        }

        if (total % 10 == 0)
        {
            if (flag == 1)
            {
                printf("AMEX\n");
            }
            if (flag == 2)
            {
                printf("MASTERCARD\n");
            }
            if (flag == 3)
            {
                printf("VISA\n");
            }
            if (flag == 0)
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    return 0;
}

int getSize(long unsigned int number)
{
    if (number > 0)
    {
        number = number / 10;
        size++;
        getSize(number);
    }

    return 0;
}

int luhnsAlgorithm(long unsigned int number)
{
    int total = 0;
    number = number / 10;
    while (number > 0)
    {
        int aux = (number % 10) * 2;

        if (aux > 9)
        {
            total = (aux % 10) + total;
            total = (aux / 10) + total;
        }
        else
        {
            total = aux + total;
        }

        number = number / 100;
    }
    return total;
}

File: pset6/sentimental-credit/credit.py
import re

def luhn_algorithm(card_number):
    total = 0
    reversed_digits = [int(digit) for digit in card_number[::-1]]

    for i, num in enumerate(reversed_digits):
        if i % 2 == 1:
            num *= 2
            if num > 9:
                num -= 9
        total += num

    return total % 10 == 0

def get_type(card_number):
    if not luhn_algorithm(card_number):
        return "INVALID"
    elif re.match(r"^3[47][0-9]{13}$", card_number):
        return "AMEX"
    elif re.match(r"^5[1-5][0-9]{14}$", card_number):
        return "MASTERCARD"
    elif re.match(r"^4[0-9]{12}(?:[0-9]{3})?$", card_number):
        return "VISA"
    return "INVALID"

card_number = input("Number: ").strip()
print(get_type(card_number))
